diff --git a/projects/openttd_vs100.vcxproj b/projects/openttd_vs100.vcxproj
index 1c3c5fa..026d8f4 100644
--- a/projects/openttd_vs100.vcxproj
+++ b/projects/openttd_vs100.vcxproj
@@ -458,6 +458,7 @@
     <ClInclude Include="..\src\landscape.h" />
     <ClInclude Include="..\src\landscape_type.h" />
     <ClInclude Include="..\src\language.h" />
+    <ClInclude Include="..\src\linkgraph_gui.h" />
     <ClInclude Include="..\src\linkgraph\demands.h" />
     <ClInclude Include="..\src\linkgraph\flowmapper.h" />
     <ClInclude Include="..\src\linkgraph\linkgraph.h" />
@@ -658,6 +659,7 @@
     <ClCompile Include="..\src\highscore_gui.cpp" />
     <ClCompile Include="..\src\industry_gui.cpp" />
     <ClCompile Include="..\src\intro_gui.cpp" />
+    <ClCompile Include="..\src\linkgraph_gui.cpp" />
     <ClCompile Include="..\src\main_gui.cpp" />
     <ClCompile Include="..\src\misc_gui.cpp" />
     <ClCompile Include="..\src\music_gui.cpp" />
diff --git a/projects/openttd_vs100.vcxproj.filters b/projects/openttd_vs100.vcxproj.filters
index e4ca8a4..fd25049 100644
--- a/projects/openttd_vs100.vcxproj.filters
+++ b/projects/openttd_vs100.vcxproj.filters
@@ -594,6 +594,9 @@
     <ClInclude Include="..\src\language.h">
       <Filter>Header Files</Filter>
     </ClInclude>
+    <ClInclude Include="..\src\linkgraph_gui.h">
+      <Filter>Header Files</Filter>
+    </ClInclude>
     <ClInclude Include="..\src\linkgraph\demands.h">
       <Filter>Header Files</Filter>
     </ClInclude>
@@ -1194,6 +1197,9 @@
     <ClCompile Include="..\src\intro_gui.cpp">
       <Filter>GUI Source Code</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\linkgraph_gui.cpp">
+      <Filter>GUI Source Code</Filter>
+    </ClCompile>
     <ClCompile Include="..\src\main_gui.cpp">
       <Filter>GUI Source Code</Filter>
     </ClCompile>
diff --git a/projects/openttd_vs80.vcproj b/projects/openttd_vs80.vcproj
index e80152e..d6552fb 100644
--- a/projects/openttd_vs80.vcproj
+++ b/projects/openttd_vs80.vcproj
@@ -1107,6 +1107,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\linkgraph_gui.h"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\linkgraph\demands.h"
 				>
 			</File>
@@ -1915,6 +1919,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\linkgraph_gui.cpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\main_gui.cpp"
 				>
 			</File>
diff --git a/projects/openttd_vs90.vcproj b/projects/openttd_vs90.vcproj
index 49fa4e6..b0e21f1 100644
--- a/projects/openttd_vs90.vcproj
+++ b/projects/openttd_vs90.vcproj
@@ -1104,6 +1104,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\linkgraph_gui.h"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\linkgraph\demands.h"
 				>
 			</File>
@@ -1912,6 +1916,10 @@
 				>
 			</File>
 			<File
+				RelativePath=".\..\src\linkgraph_gui.cpp"
+				>
+			</File>
+			<File
 				RelativePath=".\..\src\main_gui.cpp"
 				>
 			</File>
diff --git a/source.list b/source.list
index 0b0e43e..ee94698 100644
--- a/source.list
+++ b/source.list
@@ -191,6 +191,7 @@ ini_type.h
 landscape.h
 landscape_type.h
 language.h
+linkgraph_gui.h
 linkgraph/demands.h
 linkgraph/flowmapper.h
 linkgraph/linkgraph.h
@@ -410,6 +411,7 @@ group_gui.cpp
 highscore_gui.cpp
 industry_gui.cpp
 intro_gui.cpp
+linkgraph_gui.cpp
 main_gui.cpp
 misc_gui.cpp
 music_gui.cpp
diff --git a/src/linkgraph_gui.cpp b/src/linkgraph_gui.cpp
new file mode 100644
index 0000000..a6db017
--- /dev/null
+++ b/src/linkgraph_gui.cpp
@@ -0,0 +1,189 @@
+/* $Id$ */
+
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file linkgraph_gui.cpp Implementation of linkgraph overlay GUI. */
+
+#include "stdafx.h"
+#include "station_base.h"
+#include "widget_type.h"
+#include "window_gui.h"
+#include "linkgraph_gui.h"
+
+template<class Twindow>
+void LinkGraphOverlay<Twindow>::Draw() const
+{
+	std::set<StationID> seen_stations;
+	std::set<std::pair<StationID, StationID> > seen_links;
+
+	const Station *sta;
+	FOR_ALL_STATIONS(sta) {
+		/* Show links between own stations or "neutral" ones like oilrigs.*/
+		if (!HasBit(this->company_mask, sta->owner)) continue;
+		CargoID c;
+		FOR_EACH_SET_CARGO_ID(c, this->cargo_mask) {
+			if (!CargoSpec::Get(c)->IsValid()) continue;
+
+			const LinkStatMap &links = sta->goods[c].link_stats;
+			for (LinkStatMap::const_iterator i = links.begin(); i != links.end(); ++i) {
+				StationID from = sta->index;
+				StationID to = i->first;
+				if (Station::IsValidID(to) && seen_stations.find(to) == seen_stations.end()) {
+					const Station *stb = Station::Get(to);
+
+					if (!HasBit(this->company_mask, stb->owner)) continue;
+					if (sta->rect.IsEmpty() || stb->rect.IsEmpty()) continue;
+					if (seen_links.find(std::make_pair(to, from)) != seen_links.end()) continue;
+
+					Point pta = this->window->GetStationMiddle(sta);
+					Point ptb = this->window->GetStationMiddle(stb);
+					if (!this->IsLinkVisible(pta, ptb)) continue;
+
+					this->DrawForwBackLinks(pta, sta->index, ptb, stb->index);
+					seen_stations.insert(to);
+				}
+				seen_links.insert(std::make_pair(from, to));
+			}
+		}
+		seen_stations.clear();
+	}
+
+	this->DrawStationDots();
+}
+
+template<class Twindow>
+FORCEINLINE bool LinkGraphOverlay<Twindow>::IsLinkVisible(Point pta, Point ptb) const
+{
+	const NWidgetBase *wi = static_cast<const Window *>(this->window)->GetWidget<NWidgetBase>(this->widget_id);
+	return !((pta.x < 0 && ptb.x < 0) ||
+			(pta.y < 0 && ptb.y < 0) ||
+			(pta.x > (int)wi->current_x && ptb.x > (int)wi->current_x) ||
+			(pta.y > (int)wi->current_y && ptb.y > (int)wi->current_y));
+}
+
+template<class Twindow>
+void LinkGraphOverlay<Twindow>::AddLinks(StationID sta, StationID stb, LinkProperties &cargo) const
+{
+	CargoID c;
+	FOR_EACH_SET_CARGO_ID(c, this->cargo_mask) {
+		if (!CargoSpec::Get(c)->IsValid()) continue;
+		GoodsEntry &ge = Station::Get(sta)->goods[c];
+		FlowStat sum_flows = ge.GetSumFlowVia(stb);
+		const LinkStatMap &ls_map = ge.link_stats;
+		LinkStatMap::const_iterator i = ls_map.find(stb);
+		if (i != ls_map.end()) {
+			const LinkStat &link_stat = i->second;
+			this->AddStats(link_stat, sum_flows, cargo);
+		}
+	}
+}
+
+template<class Twindow>
+/* static */ void LinkGraphOverlay<Twindow>::AddStats(const LinkStat &orig_link, const FlowStat &orig_flow, LinkProperties &cargo)
+{
+	uint new_cap = orig_link.Capacity();
+	uint new_usg = orig_link.Usage();
+	uint new_plan = orig_flow.Planned();
+
+	/* multiply the numbers by 32 in order to avoid comparing to 0 too often. */
+	if (cargo.capacity == 0 ||
+			max(cargo.usage, cargo.planned) * 32 / (cargo.capacity + 1) < max(new_usg, new_plan) * 32 / (new_cap + 1)) {
+		cargo.capacity = new_cap;
+		cargo.usage = new_usg;
+		cargo.planned = new_plan;
+	}
+}
+
+template<class Twindow>
+void LinkGraphOverlay<Twindow>::DrawForwBackLinks(Point pta, StationID sta, Point ptb, StationID stb) const
+{
+	LinkProperties forward, backward;
+	this->AddLinks(sta, stb, forward);
+	this->AddLinks(stb, sta, backward);
+	GfxDrawLine(pta.x, pta.y, ptb.x, ptb.y, _colour_gradient[COLOUR_GREY][1]);
+	this->DrawContent(pta, ptb, forward);
+	this->DrawContent(ptb, pta, backward);
+
+}
+
+template<class Twindow>
+/* static */ void LinkGraphOverlay<Twindow>::DrawContent(Point pta, Point ptb, LinkProperties &cargo)
+{
+	if (cargo.capacity <= 0) return;
+	int direction_y = (pta.x < ptb.x ? 1 : -1);
+	int direction_x = (pta.y > ptb.y ? 1 : -1);;
+
+	uint usage_or_plan = min(cargo.capacity * 2, max(cargo.usage, cargo.planned));
+	int colour = LinkGraphOverlay<Twindow>::LINK_COLOURS[usage_or_plan * lengthof(LinkGraphOverlay<Twindow>::LINK_COLOURS) / (cargo.capacity * 2 + 1)];
+	GfxDrawLine(pta.x + direction_x, pta.y, ptb.x + direction_x, ptb.y, colour);
+	GfxDrawLine(pta.x, pta.y + direction_y, ptb.x, ptb.y + direction_y, colour);
+}
+
+/**
+ * Draw dots for stations into the smallmap. The dots' sizes are determined by the amount of
+ * cargo produced there, their colours by the type of cargo produced.
+ */
+template<class Twindow>
+void LinkGraphOverlay<Twindow>::DrawStationDots() const
+{
+	const Station *st;
+	FOR_ALL_STATIONS(st) {
+		if (!HasBit(this->company_mask, st->owner) || st->rect.IsEmpty()) continue;
+		Point pt = this->window->GetStationMiddle(st);
+		const NWidgetBase *wi = static_cast<const Window *>(this->window)->GetWidget<NWidgetBase>(this->widget_id);
+		if (pt.x < 0 || pt.y < 0 || pt.x > (int)wi->current_x || pt.y > (int)wi->current_y) continue;
+
+		/* Add up cargo supplied for each selected cargo type */
+		uint q = 0;
+		int colour = 0;
+		int numCargos = 0;
+		CargoID c;
+		FOR_EACH_SET_CARGO_ID(c, this->cargo_mask) {
+			if (!CargoSpec::Get(c)->IsValid()) continue;
+			uint supply = st->goods[c].supply;
+			if (supply > 0) {
+				q += supply;
+				colour += CargoSpec::Get(c)->legend_colour;
+				++numCargos;
+			}
+		}
+		if (numCargos > 1) colour /= numCargos;
+
+		uint r = 1;
+		if (q >= 20) r++;
+		if (q >= 90) r++;
+		if (q >= 160) r++;
+
+		LinkGraphOverlay::DrawVertex(pt.x, pt.y, r, colour, _colour_gradient[COLOUR_GREY][3]);
+	}
+}
+
+/**
+ * Draw a square symbolizing a producer of cargo.
+ * @param x the x coordinate of the middle of the vertex
+ * @param y the y coordinate of the middle of the vertex
+ * @param size the x and y extend of the vertex
+ * @param colour the colour with which the vertex will be filled
+ * @param border_colour the colour for the border of the vertex
+ */
+template<class Twindow>
+/* static */ void LinkGraphOverlay<Twindow>::DrawVertex(int x, int y, int size, int colour, int border_colour)
+{
+	size--;
+	int w1 = size / 2;
+	int w2 = size / 2 + size % 2;
+
+	GfxFillRect(x - w1, y - w1, x + w2, y + w2, colour);
+
+	w1++;
+	w2++;
+	GfxDrawLine(x - w1, y - w1, x + w2, y - w1, border_colour);
+	GfxDrawLine(x - w1, y + w2, x + w2, y + w2, border_colour);
+	GfxDrawLine(x - w1, y - w1, x - w1, y + w2, border_colour);
+	GfxDrawLine(x + w2, y - w1, x + w2, y + w2, border_colour);
+}
diff --git a/src/linkgraph_gui.h b/src/linkgraph_gui.h
new file mode 100644
index 0000000..803afbc
--- /dev/null
+++ b/src/linkgraph_gui.h
@@ -0,0 +1,67 @@
+/* $Id$ */
+
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file linkgraph_gui.h Declaration of linkgraph overlay GUI. */
+
+#ifndef LINKGRAPH_GUI_H_
+#define LINKGRAPH_GUI_H_
+
+#include "company_func.h"
+
+/**
+ * Properties of a link between two stations.
+ */
+struct LinkProperties {
+	LinkProperties() { this->Clear(); }
+
+	void Clear() { this->capacity = this->usage = this->planned = 0; }
+
+	uint capacity;
+	uint usage;
+	uint planned;
+};
+
+/**
+ * Handles drawing of links into some window.
+ * @tparam Twindow window type to be drawn into. Must provide methods:
+ * - Point GetStationMiddle(Station *st) const
+ */
+template<class Twindow>
+class LinkGraphOverlay {
+public:
+	static const uint8 LINK_COLOURS[];
+
+	/**
+	 * Create a link graph overlay for the specified window.
+	 * @param w Window to be drawn into.
+	 */
+	LinkGraphOverlay(const Twindow *w, uint wid, uint32 cargo_mask = 0xFFFF,
+			uint32 company_mask = (1 << _local_company) | INVALID_COMPANY) :
+			window(w), widget_id(wid), cargo_mask(cargo_mask), company_mask(company_mask)
+	{}
+
+	void Draw() const;
+
+protected:
+	const Twindow *window; ///< Window to be drawn into.
+	uint widget_id;        ///< ID of widget in window to be drawn into.
+	uint32 cargo_mask;     ///< Bitmask of cargos to be displayed.
+	uint32 company_mask;   ///< Bitmask of companies to be displayed.
+
+	void DrawForwBackLinks(Point pta, StationID sta, Point ptb, StationID stb) const;
+	void AddLinks(StationID sta, StationID stb, LinkProperties &cargo) const;
+	void DrawStationDots() const;
+	bool IsLinkVisible(Point pta, Point ptb) const;
+
+	static void AddStats(const LinkStat &orig_link, const FlowStat &orig_flow, LinkProperties &cargo);
+	static void DrawContent(Point pta, Point ptb, LinkProperties &cargo);
+	static void DrawVertex(int x, int y, int size, int colour, int border_colour);
+};
+
+#endif /* LINKGRAPH_GUI_H_ */
